File: Books/Queues/QueueList/JS/queueList.js
class Node {
    dData;
    next;

    constructor(data) {
        this.dData = data;
        this.next = null;
    }

    displayNode() {
        console.log(`{data: ${this.dData}} `);
    }
}

class LinkedList {
    #first;
    #last;

    constructor() {
        this.#first = null;
        this.#last = null;
    }

    isEmpty() {
        return this.#first === null
    }

    insertLast(data) {
        const node = new Node(data);

        this.isEmpty() ? this.#first = node : this.#last.next = node;
        this.#last = node;
    }

    deleteFirst() {
        const current = this.#first;

        if (this.isEmpty()) {
            this.#last = null;
        }
        this.#first = this.#first.next;
        return current;
    }

    displayList() {
        let current = this.#first;

        while(current !== null) {
            current.displayNode();
            current = current.next;
        }
    }
}

class QueueList {
    theQueue;

    constructor() {
        this.theQueue = new LinkedList();
    }

    push(data) {
        this.theQueue.insertLast(data)
    }

    pop() {
        this.theQueue.deleteFirst()
    }

    displayQueue() {
        console.log("Queue (start -> end): ")
        this.theQueue.displayList();
    }
}

(function QueueListApp() {
    const theQueue = new QueueList();

    theQueue.push(10);
    theQueue.push(20);
    theQueue.push(30);

    theQueue.displayQueue();

    theQueue.push(40);
    theQueue.push(50);
    theQueue.push(60);

    theQueue.displayQueue();

    theQueue.pop();
    theQueue.pop();
    theQueue.pop();
    theQueue.pop();

    theQueue.displayQueue();
})()

File: Books/JavaRLafore/Queues/Java/QueueList/queueList.java
package Books.Queues.Java.QueueList;

class QueueListApp {
    public static void main(String[] args) {

        QueueList theQueueList = new QueueList();

        theQueueList.push(10);
        theQueueList.push(20);
        theQueueList.push(30);

        theQueueList.displayQueue();

        theQueueList.push(40);
        theQueueList.push(50);
        theQueueList.push(60);

        theQueueList.displayQueue();

        theQueueList.pop();
        theQueueList.pop();
        theQueueList.pop();
        theQueueList.pop();

        theQueueList.displayQueue();
    }
}

class Link {
    public long dData;
    public Link next;

    public Link(long data) {
        dData = data;
    }

    public void displayLink() {
        System.out.print("{data: " + dData + "} ");
    }
}

class LinkedList {
    private Link first;
    private Link last;

    public LinkedList() {
        first = null;
        last = null;
    }

    public boolean isEmpty() {
        return first == null;
    }
    
    public void insertLast(long data) {
        Link newLink = new Link(data);

        if(isEmpty()) {
            first = newLink;
        } else {
            last.next = newLink;
        }
        last = newLink;
    }

    public long deleteFirst() {
        long temp = first.dData;
        if (isEmpty()) {
            last = null;
        }
        first = first.next;
        return temp;
    }

    public void displayList() {
        System.out.print("List (first --> last): ");
        Link current = first;

        while (current != null) {
            current.displayLink();
            current = current.next;
        }

        System.out.println("");
    }
}

class QueueList {
    private LinkedList theQueue;

    public QueueList() {
        theQueue = new LinkedList();
    }

    public void push(long data) {
        theQueue.insertLast(data);
    }

    public long pop() {
        return theQueue.deleteFirst();
    }

    public boolean isEmpty() {
        return theQueue.isEmpty();
    }

    public void displayQueue() {
        System.out.print("Queue (start -> end): ");
        theQueue.displayList();
    }
}
